### PR TITLE
Implement a constrained sorting gather variety

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Currently, the `sort-limit` optimizer rule is broken (respectively disabled by #10047) in the cluster when combined with fullCount. To fix this, the SortingGatherExecutor is made aware of the limit, like the ConstrainedSortExecutor is.

In addition, the `sort-limit` rule can now make the `SortingGatherExecutor` aware of the limit, even when there is no `SORT` node. This will improve the performance of a result sorted by an index in conjunction with `fullCount`, as the counting can now happen on the DBServers.

- [X] Bug-Fix for *devel-branch*
- [ ] Bug-Fix for *3.5* (**must include #10047**)
- [X] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as `shell_server_aql`, specifically:
- `aql-queries-optimizer-sort-limit-cluster.js` as added by https://github.com/arangodb/arangodb/pull/10047, and
- `aql-profiler.js` as well as
- `aql-profiler-cluster.js`.

### Documentation

- [x] Added a *Changelog Entry*

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6392/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6400/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6424/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6433/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6441/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6447/

**Comment by goedderz** [comment]
The last remaining failures look unrelated. Pulled devel again, retry:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6455/

**Comment by goedderz** [comment]
The failing test is unrelated and being investigated.

File: arangod/Aql/DependencyProxy.cpp [hunk 1]
```diff
@@ -142,6 +142,41 @@ DependencyProxy<passBlocksThrough>::fetchBlockForDependency(size_t dependency, s
   return {state, block};
 }
 
+template <bool allowBlockPassthrough>
+std::pair<ExecutionState, size_t> DependencyProxy<allowBlockPassthrough>::skipSomeForDependency(
```
Comment (mchacki) [inline arangod/Aql/DependencyProxy.cpp]: yes this was missing \o/

File: arangod/Aql/OptimizerRule.h [hunk 2]
```diff
@@ -285,6 +282,9 @@ struct OptimizerRule {
     // push collect operations to the db servers
     collectInClusterRule,
 
+    // make sort node aware of subsequent limit statements for internal optimizations
+    applySortLimitRule,
```
Comment (mchacki) [inline arangod/Aql/OptimizerRule.h]:
```suggestion
    // This is required to run AFTER the cluster nodes have been inserted. As it modifies the GatherNode
    applySortLimitRule,
```

**Pull request merged**